README.md
# verilog
Verilog Hardware Description Language Project

This repository consists of projects I have worked on using Verilog HDL during my time at the University of Toronto.
The projects in this repository are based on the following courses:

• ECE 241 F – Digital Systems (Fall 2023), Language: Verilog HDL

Currently this repository contains only my final project for ECE 241.


ece241-final/changelog.txt
 UPDATED 11/30 3:30 AM
 signals are outputting, however we are not moving on from STOP_HORIZONTAL state.

 UPDATED 11/30 5:30 PM
 display_counter is working normally, may be an issue with clock crossings
 added hold variable in display_counter to keep done on for longer, which triggered the HEXs to start working

 UPDATED 11/30 7:00 PM
 changed as many posedge Clock to be @ (*), still not outputting changes, and CounterValue from display_counter is not being passed back

 UPDATE 12/01 11:00 AM
 changed all clocks to CLOCK_50 and somehow that worked, not sure how to incorporate rate divider now
 Try sending CLOCK_50 to everything and using std_clock as an enable signal -- always @ (CLOCK_50), if(std_clock).

 UPDATE 12/01 3:00PM
 noticed must implement seperate colours for horizontal and vertical
 to try: reconnecting all the different enable signals
 check display_counter module as the counters are not outputing correctly (only starting from 3 and counting down at every cycle)

 UPDATE 12/01 4:30 PM
 half working, display_counter highest is not updating at all for different states
 highest and 0 are being combined into one clock edge which is causing funky issues

 UPDATE 12/02 1:00 PM
 fixed hex-decoder, all #s from 0-25 should now be outputting the right numbers
 added buffer for done signal, countdown now counting from 31 every time (no chance to set highest still)
 to fix logic in display_counter; not executing on edge of std_clock
 Made copy draft1.v to test display colours

 UPDATE 12/02 2:00 PM
 highest values are now being inputted but not at the correct cycle, might have to put highest calculation inside when count reaches 0
 changed assignments to highest to be blocking assignments, seems to be working correctly

 UPDATE 12/02 3:30 PM
 Enable being held on for more than 1 posedge CLOCK_50 is skipping to the next cycle too fast
 going to introduce hold states into FSM that waits for enable signal to go low before proceeding
 hold states work, but signals do not stay on for hold state, try enabling same signals during hold states

 UPDATE 12/03 12:00 AM
 added hold states which seem to fix the signals, only would have to check if red is working as expected, or if it also needs a wait state

 UPDATE 12/03 3:30 PM
 placing a hold on testing the original code, made a copy as draft2.v
 starting to implement the PS2 connections and VGA connections to the board - display them on the LEDRs (new hex decoder)
 and use the swicthes as replacement R-G-B PS2 keyboard inputs
 remaking entire lights_datapath module

PARTNER-INTEGRATION UPDATE
 PS2-KEYBOARD WILL SEND A 3-BIT INPUT (we can truncate and use each as an R/G/B signal), 
    SIMULATE USING BUTTONS KEY0, KEY1, KEY2:
    variable input_signal
                                                go to:
    3b’000 - vertical green, horizontal red     GV_HOLD
    001 - vertical yellow, horizontal red       YV_HOLD
    010 - vertical red, horizontal green        GH_HOLD
    011 - vertical red, horizontal yellow       YH_HOLD
    100 - all red, vertical nxt                 SV_HOLD                               
    101 -          horizontal next              SH_HOLD

 From Testing:
 countdown doesn't display 0, maybe I want to do that?
 mid-reset only goes to the beginning of the current state
 green light executes before walk signal, which is fine - but we can introduce a walk-before-light for complexity controlled by SW[2]

 UPDATE 12/04 11:00 AM
 implemented forced signals, ModelSim shows they seem to be jumping around to different states properly
 fixed Reset

 UPDATE 12/04 11:00 PM
 added 3-bit vga_signal according to above mappings (same for input_signal)

 UPDATE 12/05 3:00 PM
 integration state - integrated parts with PS2-Keyboard and VGA code
 original mapped KEY[3:1] to represent 3-bit input signal (as above), KEY[0] changed to reset (instead of SW[9])
 debugged common anode signals when transferring KEY[3:1] to PS2 Keyboard received signals

 PRESENTATION – 12/06 10:00 AM


hex-decoder.v
module hex_decoder(c, display);
    input [3:0] c;
    output [6:0] display;
    assign display[0] = ~((c[3]|c[2]|c[1]|~c[0]) & (c[3]|~c[2]|c[1]|c[0]) & (~c[3]|c[2]|~c[1]|~c[0]) & (~c[3]|~c[2]|c[1]|~c[0]));
    assign display[1] = ~((c[3]|~c[2]|c[1]|~c[0]) & (c[3]|~c[2]|~c[1]|c[0]) & (~c[3]|c[2]|~c[1]|~c[0]) & (~c[3]|~c[2]|c[1]|c[0]) & (~c[3]|~c[2]|~c[1]|c[0]) & (~c[3]|~c[2]|~c[1]|~c[0]));
    assign display[2] = ~((c[3]|c[2]|~c[1]|c[0]) & (~c[3]|~c[2]|c[1]|c[0]) & (~c[3]|~c[2]|~c[1]|c[0]) & (~c[3]|~c[2]|~c[1]|~c[0]));
    assign display[3] = ~((c[3]|c[2]|c[1]|~c[0]) & (c[3]|~c[2]|c[1]|c[0]) & (c[3]|~c[2]|~c[1]|~c[0]) & (~c[3]|c[2]|~c[1]|c[0]) & (~c[3]|~c[2]|~c[1]|~c[0]));
    assign display[4] = ~((c[3]|c[2]|c[1]|~c[0]) & (c[3]|c[2]|~c[1]|~c[0]) & (c[3]|~c[2]|c[1]|c[0]) & (c[3]|~c[2]|c[1]|~c[0]) & (c[3]|~c[2]|~c[1]|~c[0]) & (~c[3]|c[2]|c[1]|~c[0]));
    assign display[5] = ~((c[3]|c[2]|c[1]|~c[0]) & (c[3]|c[2]|~c[1]|c[0]) & (c[3]|c[2]|~c[1]|~c[0]) & (c[3]|~c[2]|~c[1]|~c[0]) & (~c[3]|~c[2]|c[1]|~c[0]));
    assign display[6] = ~((c[3]|c[2]|c[1]|c[0]) & (c[3]|c[2]|c[1]|~c[0]) & (c[3]|~c[2]|~c[1]|~c[0]) & (~c[3]|~c[2]|c[1]|c[0]));
endmodule


ece241-final/project-main.v
// Author: Tristan Seto
// Date: 2023-12-06

// This file contains the top-level module for the complete project.

module main(
    input CLOCK_50,
    input [9:0] SW,
    input [3:0] KEY,
    inout PS2_CLK, PS2_DAT,
    output [3:0] LEDR,
    output [6:0] HEX3, HEX2, HEX1, HEX0,
    output VGA_CLK, VGA_HS, VGA_VS, VGA_BLANK_N, VGA_SYNC_N,
    output [7:0] VGA_R, VGA_G, VGA_B);

    parameter CLOCK_FREQUENCY = 50000000;

    wire std_clock; // standard clock will count per 1 second
    wire forced_reset;
    wire input_signal;
    wire Enable;
    wire green_hor, yellow_hor, red_hor;
    wire green_ver, yellow_ver, red_ver;
    wire walk_hor, countdown_hor, stop_hor;
    wire walk_ver, countdown_ver, stop_ver;
    wire [2:0] colour_hor, colour_ver;
    wire [2:0] vga_signal;
    wire [4:0] CounterValue;

    assign forced_reset = (input_signal != 3'b111);

    // Instatiate a clock that ticks every one second
    rate_divider #(CLOCK_FREQUENCY) standard_clock(.ClockIn(CLOCK_50), .Reset(~KEY[0]), .Enable(std_clock));

    // PS2
    PS2_Demo ps2(.CLOCK_50(CLOCK_50), .KEY(~KEY[0]), .PS2_CLK(PS2_CLK), .PS2_DAT(PS2_DAT), .output_3bits(input_signal));

    // Call display counter module (this module will use a half-second clock for the blinkers)
    display_counter u0(
        .Clock(CLOCK_50), .Reset(~KEY[0] || forced_reset), .Enable(std_clock),
        .red(red_hor || red_ver), .yellow(yellow_hor || yellow_ver), .green(walk_hor || walk_ver), .countdown(countdown_hor || countdown_ver),
        .num_of_cars(SW[1:0]),
        //.Clk(half_clock), // comment out
        .CounterValue(CounterValue),        
        .done(Enable)//, .display(display)
        );
    
    FSM u1(.Clock(CLOCK_50), .Reset(~KEY[0]), .Enable(Enable),
        .input_signal(input_signal), // change when PS2 input comes in
        .green_hor(green_hor), .yellow_hor(yellow_hor), .red_hor(red_hor),
        .green_ver(green_ver), .yellow_ver(yellow_ver), .red_ver(red_ver),
        .walk_hor(walk_hor), .countdown_hor(countdown_hor), .stop_hor(stop_hor),
        .walk_ver(walk_ver), .countdown_ver(countdown_ver), .stop_ver(stop_ver)
    );

    lights_datapath u2(
        .Clock(CLOCK_50), .Reset(~KEY[0]),
        .green_hor(green_hor), .yellow_hor(yellow_hor), .red_hor(red_hor),
        .green_ver(green_ver), .yellow_ver(yellow_ver), .red_ver(red_ver),
        .colour_hor(colour_hor), .colour_ver(colour_ver), .out_signal(vga_signal)
    );

    crosswalk_datapath #(CLOCK_FREQUENCY) u3(
        .ClockIn(CLOCK_50), .Reset(~KEY[0]),
        .walk_hor(walk_hor), .countdown_hor(countdown_hor), .stop_hor(stop_hor),
        .walk_ver(walk_ver), .countdown_ver(countdown_ver), .stop_ver(stop_ver),
        .go_h(LEDR[3]), .stop_h(LEDR[2]), .go_v(LEDR[1]), .stop_v(LEDR[0])
    );

    countdown_hex_decoder hex(.Enable(countdown_hor || countdown_ver), .c(CounterValue), .dispR(HEX0), .dispL(HEX1));

    colour_hex_decoder light_horizontal(.c(colour_hor), .display(HEX3));
    colour_hex_decoder light_vertical(.c(colour_ver), .display(HEX2));

    fill vga_map(.CLOCK_50(CLOCK_50),
		.KEY(KEY),
		.my_3_bit_input(vga_signal),
		.VGA_CLK(VGA_CLK),   						//	VGA Clock
		.VGA_HS(VGA_HS),							//	VGA H_SYNC
		.VGA_VS(VGA_VS),							//	VGA V_SYNC
		.VGA_BLANK_N(VGA_BLANK_N),					//	VGA BLANK
		.VGA_SYNC_N(VGA_SYNC_N),					//	VGA SYNC
		.VGA_R(VGA_R),   						//	VGA Red[9:0]
		.VGA_G(VGA_G),	 						//	VGA Green[9:0]
		.VGA_B(VGA_B)   						//	VGA Blue[9:0]
	);

endmodule


ece241-final/project-addons.v
// Author: Tristan Seto
// Date: 2023-12-06
// Description: This files contains all the additional parts (add-ons) to my ECE 241 F final project
// that are not a part of the main functionality, but act as helpers to instantiate parts of the project.
// This includes rate divider modules and hex decoder modules.

// Rate Divider
module rate_divider
#(parameter CLOCK_FREQUENCY = 50000000) (input ClockIn, input Reset, output reg Enable);
    // may want to try do a half clock frequency to do a blinking light
    
    wire [31:0] highest;
    reg [31:0] countdown;
    assign highest = CLOCK_FREQUENCY - 1;

    always @ (posedge ClockIn)
        begin
            if(Reset || countdown == 32'b0) countdown <= highest;
            else countdown <= countdown - 1;
        end

    always @ (*)
    begin
        if(countdown != 32'b0) Enable <= 1'b0;
        else Enable <= 1'b1;
    end
    
endmodule


// Hex Decoder for the Countdown
module countdown_hex_decoder(Enable, c, dispR, dispL);
    // only counts values from 1 to 25; @ 0 it will display nothing
    input Enable;
    input [4:0] c;
    output reg [6:0] dispR, dispL;

    // If the # is greater than 25 (or 0), do not display any numbers
    always @ (*)
    begin
      if(!Enable || c > 6'd25) begin // Display will be blank if the input number is reset or is greater than 25
            dispR = 7'b1111111;
            dispL = 7'b1111111;
        end else begin
            // Else display the numbers counting down to zero

        // Right Hand Side (PoS active low, convert to active high)
        dispR[0] = ~((c[4]|c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|c[3]|~c[2]|c[1]|c[0]) & (c[4]|~c[3]|c[2]|~c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|~c[1]|c[0]) & (~c[4]|c[3]|~c[2]|c[1]|~c[0]) & (~c[4]|~c[3]|c[2]|c[1]|c[0]));
        dispR[1] = ~((c[4]|c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|~c[2]|c[1]|~c[0]) & (c[4]|c[3]|~c[2]|~c[1]|c[0]) & (c[4]|~c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|c[2]|c[1]|c[0]) & (~c[4]|~c[3]|c[2]|c[1]|~c[0]));
        dispR[2] = ~((c[4]|c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|c[2]|~c[1]|c[0]) & (c[4]|~c[3]|~c[2]|c[1]|c[0]) & (~c[4]|c[3]|~c[2]|~c[1]|c[0]));
        dispR[3] = ~((c[4]|c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|c[3]|~c[2]|c[1]|c[0]) & (c[4]|~c[3]|c[2]|~c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|~c[1]|c[0]) & (~c[4]|c[3]|~c[2]|c[1]|~c[0]) & (~c[4]|~c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|c[2]|c[1]|~c[0]));
        dispR[4] = ~((c[4]|c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|c[3]|~c[2]|c[1]|c[0]) & (c[4]|~c[3]|c[2]|~c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|~c[1]|c[0]) & (~c[4]|c[3]|~c[2]|c[1]|~c[0]) & (~c[4]|~c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|c[3]|~c[2]|c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|~c[3]|c[2]|c[1]|~c[0]) & (c[4]|c[3]|c[2]|~c[1]|~c[0]) & (c[4]|~c[3]|c[2]|c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|c[1]|~c[0]) & (~c[4]|c[3]|c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|~c[2]|~c[1]|~c[0]));
        dispR[5] = ~((c[4]|c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|~c[3]|c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|~c[2]|c[1]|~c[0]) & (c[4]|c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|c[3]|c[2]|~c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|c[1]|~c[0]) & (~c[4]|c[3]|~c[2]|~c[1]|~c[0]) & (c[4]|c[3]|c[2]|~c[1]|c[0]) & (c[4]|~c[3]|~c[2]|c[1]|c[0]) & (~c[4]|c[3]|~c[2]|~c[1]|c[0]));
        dispR[6] = ~((c[4]|c[3]|c[2]|c[1]|c[0]) & (c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|~c[3]|c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|~c[2]|c[1]|~c[0]) & (c[4]|c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|c[2]|c[1]|~c[0]) & (c[4]|~c[3]|c[2]|~c[1]|c[0]) & (~c[4]|c[3]|~c[2]|c[1]|c[0]));

        // Left Hand Side (PoS active high)
        dispL[0] = ((~c[4]|c[3]|~c[2]|c[1]|c[0]) & (~c[4]|c[3]|~c[2]|c[1]|~c[0]) & (~c[4]|c[3]|~c[2]|~c[1]|c[0]) & (~c[4]|c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|~c[3]|c[2]|c[1]|c[0]) & (~c[4]|~c[3]|c[2]|c[1]|~c[0]));
        dispL[2] = ((c[4]|~c[3]|c[2]|~c[1]|c[0]) & (c[4]|~c[3]|c[2]|~c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|c[1]|c[0]) & (c[4]|~c[3]|~c[2]|c[1]|~c[0]) & (c[4]|~c[3]|~c[2]|~c[1]|c[0]) & (c[4]|~c[3]|~c[2]|~c[1]|~c[0]) & (~c[4]|c[3]|c[2]|c[1]|c[0]) & (~c[4]|c[3]|c[2]|c[1]|~c[0]) & (~c[4]|c[3]|c[2]|~c[1]|c[0]) & (~c[4]|c[3]|c[2]|~c[1]|~c[0]));
        dispL[1] = dispL[0] & dispL[2];
        dispL[3] = dispL[0];
        dispL[4] = dispL[0];
        dispL[5] = 1'b1;
        dispL[6] = dispL[0];

        end
    end
endmodule

// Hex Decoder for the Colours:
module colour_hex_decoder(c, display);
    input [2:0] c;
    output [6:0] display;

    assign display[0] = (c[2]&c[1]&c[0]) | (c[2]&c[1]&~c[0]) | (~c[2]&c[1]&c[0]);
    assign display[1] = (~c[2]&c[1]&~c[0]) | (~c[2]&~c[1]&c[0]) | (c[2]&c[1]&c[0]);
    assign display[2] = (c[2]&~c[1]&c[0]) | (c[2]&c[1]&c[0]);
    assign display[3] = (~c[2]&~c[1]&~c[0]) | (c[2]&c[1]&~c[0]) | (~c[2]&c[1]&c[0]) | (c[2]&c[1]&c[0]);
    assign display[4] = ~(c[2]&~c[1]&c[0]);
    assign display[5] = ~((~c[2]&c[1]&~c[0]) | (~c[2]&c[1]&c[0]));
    assign display[6] = (~c[2]&~c[1]&~c[0]) | (c[2]&c[1]&~c[0]) | (c[2]&c[1]&c[0]);

  // As the design is common anode, the colour mappings are as follows:
  // RED:     FF0000 (255-RGB) = 100 (1-RGB) = "3" (011)
  // YELLOW:  FFFF00 (255-RGB) = 110 (1-RGB) = "1" (001)
  // GREEN:   00FF00 (255-RGB) = 010 (1-RGB) = "5" (101)
endmodule


ece241-final/project-fsm.v
// Author: Tristan Seto
// Date: 2023-12-06
// Description: This file contains the project's Finite State Machine (FSM) control path and data paths for both
// traffic light signals and pedestrian crossing signals. 
// The countdown/counter module that counts the time between states is also declared and defined in this file.


// Finite State Machine (FSM) control path
module FSM(
    input Clock,
    input Reset, Enable,
    input [2:0] input_signal,
    output reg green_hor, yellow_hor, red_hor,
    output reg green_ver, yellow_ver, red_ver,
    output reg walk_hor, countdown_hor, stop_hor,
    output reg walk_ver, countdown_ver, stop_ver
    );

    reg [3:0] current_state, next_state;

    localparam  STOP_HORIZONTAL           = 4'd0,
                SH_HOLD                   = 4'd1,
                GO_HORIZONTAL             = 4'd2,
                GH_HOLD                   = 4'd3,
                COUNTDOWN_HORIZONTAL      = 4'd4,
                CH_HOLD                   = 4'd5,
                YELLOW_HORIZONTAL         = 4'd6,
                YH_HOLD                   = 4'd7,     
                STOP_VERTICAL             = 4'd8,
                SV_HOLD                   = 4'd9,
                GO_VERTICAL               = 4'd10,
                GV_HOLD                   = 4'd11,
                COUNTDOWN_VERTICAL        = 4'd12,
                CV_HOLD                   = 4'd13,
                YELLOW_VERTICAL           = 4'd14,
                YV_HOLD                   = 4'd15;  

    // Next state logic

    // check after every posedge of Clock
    // States should only update when the Enable signal is sent from display_counter
    always @ (*) 
    if(Reset) next_state = SH_HOLD;
    else
    begin: state_table
    if(input_signal == 3'b000)      next_state = GV_HOLD;
    else if(input_signal == 3'b001) next_state = YV_HOLD;
    else if(input_signal == 3'b010) next_state = GH_HOLD;
    else if(input_signal == 3'b011) next_state = YH_HOLD;
    else if(input_signal == 3'b100) next_state = SV_HOLD;
    else if(input_signal == 3'b101) next_state = SH_HOLD;
    else begin
            case (current_state)
                STOP_HORIZONTAL: 
                    next_state = Enable ? STOP_HORIZONTAL : SH_HOLD;
                SH_HOLD: 
                    next_state = Enable ? GO_HORIZONTAL : SH_HOLD;
                GO_HORIZONTAL:
                    next_state = Enable ? GO_HORIZONTAL : GH_HOLD;
                GH_HOLD:
                    next_state = Enable ? COUNTDOWN_HORIZONTAL : GH_HOLD;
                COUNTDOWN_HORIZONTAL:
                    next_state = Enable ? COUNTDOWN_HORIZONTAL : CH_HOLD;
                CH_HOLD:
                    next_state = Enable ? YELLOW_HORIZONTAL : CH_HOLD;
                YELLOW_HORIZONTAL:
                    next_state = Enable ? YELLOW_HORIZONTAL : YH_HOLD;
                YH_HOLD:
                    next_state = Enable ? STOP_VERTICAL : YH_HOLD;                
                STOP_VERTICAL: 
                    next_state = Enable ? STOP_VERTICAL : SV_HOLD;
                SV_HOLD: 
                    next_state = Enable ? GO_VERTICAL : SV_HOLD;
                GO_VERTICAL:
                    next_state = Enable ? GO_VERTICAL : GV_HOLD;
                GV_HOLD:
                    next_state = Enable ? COUNTDOWN_VERTICAL : GV_HOLD;
                COUNTDOWN_VERTICAL:
                    next_state = Enable ? COUNTDOWN_VERTICAL : CV_HOLD;
                CV_HOLD:
                    next_state = Enable ? YELLOW_VERTICAL : CV_HOLD;
                YELLOW_VERTICAL:
                    next_state = Enable ? YELLOW_VERTICAL : YV_HOLD;
                YV_HOLD:
                    next_state = Enable ? STOP_HORIZONTAL : YV_HOLD;
            default:     next_state = SH_HOLD;
        endcase
    end
    end // state_table

    // Output logic aka all of our datapath control signals
    always @ (posedge Clock)
    begin: enable_signals
        // By default make all our signals 0:

        // Traffic Light Signals
        green_hor = 1'b0;
        yellow_hor = 1'b0;
        red_hor = 1'b0;
        // left_hor = 1'b0;
        green_ver = 1'b0;
        yellow_ver = 1'b0;
        red_ver = 1'b0;
        // left_ver = 1'b0;

        // Pedestrian Signals
        walk_hor = 1'b0;        // Walk LED (3) ON
        countdown_hor = 1'b0;   // Stop LED (2) FLASHING 1/2 RATE OF CLOCK, ENABLE COUNTODWN
        stop_hor = 1'b0;        // Stop LED (2) ON
        walk_ver = 1'b0;        // Walk LED (1) ON
        countdown_ver = 1'b0;   // Stop LED (0) FLASHING 1/2 RATE OF CLOCK, ENABLE COUNTDOWN
        stop_ver = 1'b0;        // Stop LED (0) ON
        // Stop LEDs (2) and 0 should be on by default

        case (current_state)
            STOP_HORIZONTAL: begin // all red signals should be ON
                // horizontal signals
                red_hor = 1'b1;
                stop_hor = 1'b1;
                // vertical signals
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            SH_HOLD: begin // all red signals should be ON
                // horizontal signals
                red_hor = 1'b1;
                stop_hor = 1'b1;
                // vertical signals
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            STOP_VERTICAL: begin // all red signals should be ON
                // horizontal signals
                red_hor = 1'b1;
                stop_hor = 1'b1;
                // vertical signals
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            SV_HOLD: begin // all red signals should be ON
                // horizontal signals
                red_hor = 1'b1;
                stop_hor = 1'b1;
                // vertical signals
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            // HORIZONTAL GO STATES
            GO_HORIZONTAL: begin
                // horizontal signals set to go
                green_hor = 1'b1;
                walk_hor = 1'b1;

                // vertical signals unchanged
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            GH_HOLD: begin
                // horizontal signals set to go
                green_hor = 1'b1;
                walk_hor = 1'b1;

                // vertical signals unchanged
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            COUNTDOWN_HORIZONTAL: begin
                green_hor = 1'b1; // traffic light remains green
                countdown_hor = 1'b1; // act as countdown enable

                // vertical signals unchanged
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            CH_HOLD: begin
                green_hor = 1'b1; // traffic light remains green
                countdown_hor = 1'b1; // act as countdown enable

                // vertical signals unchanged
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            YELLOW_HORIZONTAL: begin
                yellow_hor = 1'b1; // traffic light turns yellow
                stop_hor = 1'b1; // crossing should be complete

                // vertical signals unchanged
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            YH_HOLD: begin
                yellow_hor = 1'b1; // traffic light turns yellow
                stop_hor = 1'b1; // crossing should be complete

                // vertical signals unchanged
                red_ver = 1'b1;
                stop_ver = 1'b1;
            end

            // VERTICAL GO STATES
            GO_VERTICAL: begin
                // horizontal signals remain unchanged
                red_hor = 1'b1;
                stop_hor = 1'b1;

                // vertical signals set to go
                green_ver = 1'b1;
                walk_ver = 1'b1;
            end

            GV_HOLD: begin
                // horizontal signals remain unchanged
                red_hor = 1'b1;
                stop_hor = 1'b1;

                // vertical signals set to go
                green_ver = 1'b1;
                walk_ver = 1'b1;
            end

            COUNTDOWN_VERTICAL: begin
                // horizontal signals remain unchanged
                red_hor = 1'b1;
                stop_hor = 1'b1;

                green_ver = 1'b1; // traffic light remains green
                countdown_ver = 1'b1; // act as countdown enable
            end

            CV_HOLD: begin
                // horizontal signals remain unchanged
                red_hor = 1'b1;
                stop_hor = 1'b1;

                green_ver = 1'b1; // traffic light remains green
                countdown_ver = 1'b1; // act as countdown enable
            end

            YELLOW_VERTICAL: begin
                // horizontal signals remain unchanged
                red_hor = 1'b1;
                stop_hor = 1'b1;

                yellow_ver = 1'b1; // traffic light turns yellow
                stop_ver = 1'b1; // crossing should be complete
            end

            YV_HOLD: begin
                // horizontal signals remain unchanged
                red_hor = 1'b1;
                stop_hor = 1'b1;

                yellow_ver = 1'b1; // traffic light turns yellow
                stop_ver = 1'b1; // crossing should be complete
            end

        // default:    // don't need default since we already made sure all of our outputs were assigned a value at the start of the always block
        endcase
    end // enable_signals

    // current_state registers
    always @ (posedge Clock)
    begin: state_FFs
        current_state <= next_state;
    end // state_FFS
endmodule



// Datapath for Traffic Light signals
module lights_datapath(
    input Clock,
    input Reset,
    input green_hor, yellow_hor, red_hor,
    input green_ver, yellow_ver, red_ver,
    output reg [2:0] colour_hor, colour_ver, out_signal

    );

    localparam  RED = 3'b100,
                YELLOW = 3'b110,
                GREEN = 3'b010;

    always @ (*) begin
        if(Reset) begin
            colour_hor <= RED;
            colour_ver <= RED;
        end

        if(red_hor) colour_hor <= RED;
        else if(green_hor) colour_hor <= GREEN;
        else if(yellow_hor) colour_hor <= YELLOW;

        if(red_ver) colour_ver <= RED;
        else if(green_ver) colour_ver <= GREEN;
        else if(yellow_ver) colour_ver <= YELLOW;

        // VGA SIGNALS
        if(green_ver) out_signal       <= 3'b000;
        else if(yellow_ver) out_signal <= 3'b001;
        else if(green_hor)  out_signal <= 3'b010;
        else if(yellow_hor) out_signal <= 3'b011;
        else out_signal <= 3'b100;
    end
endmodule


// Datapath for Pedestrian Crosswalk signals
module crosswalk_datapath
#(parameter CLOCK_FREQUENCY = 50000000)(
    input ClockIn,
    input Reset,
    input walk_hor, countdown_hor, stop_hor,
    input walk_ver, countdown_ver, stop_ver,
    output reg go_h, stop_h, go_v, stop_v

    );

    wire countOn;
    rate_divider #(CLOCK_FREQUENCY / 2) crosswalk_count(.ClockIn(ClockIn), .Reset(Reset), .Enable(countOn));

    always @ (posedge countOn) begin
        if(Reset) begin
            stop_h <= 1'b1;
            go_h <= 1'b0;
            stop_v <= 1'b1;
            go_v <= 1'b0;
        end
        
        // horizontal signals
        if(stop_hor) begin
            stop_h <= 1'b1;
            go_h <= 1'b0;
        end
        else if(walk_hor) begin
            stop_h <= 1'b0;
            go_h <= 1'b1;
        end
        else if(countdown_hor) begin
            stop_h <= !stop_h;
            go_h <= 1'b0;
        end

        // vertical signals
        if(stop_ver) begin
            stop_v <= 1'b1;
            go_v <= 1'b0;
        end
        else if(walk_ver) begin
            stop_v <= 1'b0;
            go_v <= 1'b1;
        end
        else if(countdown_ver) begin
            stop_v <= !stop_v;
            go_v <= 1'b0;
        end
    end
endmodule



// Display Counter
// This module sets the countdown time (from user input) and counts down the time for each state accordingly.

module display_counter(
        input Clock, Reset, Enable,
        input red, yellow, green, countdown,
        input [1:0] num_of_cars,
        output reg [4:0] CounterValue,
        output reg done
        );

    reg [5:0] highest;

    always @ (posedge Clock)
    begin
        
        // default done signal
        done <= 1'b0;
        
        // input highest value
        if(yellow) begin
            highest = 6'd3;
        end
        else if (green) begin
            case (num_of_cars)
                2'd0: highest = 5'd5;
                2'd1: highest = 5'd5;
                2'd2: highest = 5'd10;
                2'd3: highest = 5'd15;
            endcase
        end
        else if (countdown) begin
            case (num_of_cars)
                2'd0: highest = 5'd10;
                2'd1: highest = 5'd15;
                2'd2: highest = 5'd20;
                2'd3: highest = 5'd25;
            endcase
        end
        else highest = 6'd3; // default if necessary

        // actual counter
        if(Reset) CounterValue = highest;
        else if(Enable) begin 
            done <= 1'b0;
            if (CounterValue == 5'b0) CounterValue = highest;
            else CounterValue <= CounterValue - 1;
            
        end
        else if (CounterValue == 5'b0) begin
            done <= 1'b1;
        end
        else CounterValue <= CounterValue;

    end

endmodule
